### PR TITLE
Implement shake

### PR DESCRIPTION
**Pull request by razor-x**

### CODE REVIEW

File: src/lib/object.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export const shake = <T>(obj: T): Partial<T> => {
```
Comment (phpnode) [inline src/lib/object.ts:1]:
```suggestion
type Shook<T> = {[K in keyof T as T[K] extends undefined ? never : K]: T[K]}
export const shake = <T>(obj: T): Shook<T> => {
```

**Review by phpnode — APPROVED** [review]

File: src/lib/object.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export const shake = <T>(obj: T): Shook<T> => {
+  if (obj == null) return obj
+  if (typeof obj !== 'object') return obj
+
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  const initialValue = {} as Shook<T>
```
Comment (razor-x) [inline src/lib/object.ts:6]: @phpnode I can't figure out how to do this without a type assertion. 🤷🏻

Comment (phpnode) [reply src/lib/object.ts:6]: you can't, TS isn't clever enough unfortunately.

Comment (phpnode) [reply src/lib/object.ts:6]: if we were being strict we'd say that this is a `Partial<T>` here, and it's only when we return that it's `as Shook<T>`, but I don't think it matters at all

**Pull request merged**